File: src/shared/lib/image-processing/light/inversion.ts
import type { FilterProcessFn } from "../types";

export const applyInversion: FilterProcessFn = (pixels) => {
  const len = pixels.length;
  for (let i = 0; i < len; i += 4) {
    pixels[i] = 255 - pixels[i];
    pixels[i + 1] = 255 - pixels[i + 1];
    pixels[i + 2] = 255 - pixels[i + 2];
  }
  return pixels;
};

File: c/implementations/light/inversion.c
#include <stddef.h>
#include <stdint.h>

void apply_inversion(uint8_t *pixels, size_t width, size_t height) {

  size_t pixel_count = width * height * PX_SIZE;

  for (size_t i = 0; i < pixel_count; i += PX_SIZE) {
    pixels[i] = 255 - pixels[i];
    pixels[i + 1] = 255 - pixels[i + 1];
    pixels[i + 2] = 255 - pixels[i + 2];
  }
}
